sql/rpl_gtid.cc

```diff
@@ -849,9 +861,20 @@ rpl_slave_state::gtid_delete_pending(THD *thd,
 
 #ifdef WITH_WSREP
   /*
-    Updates in slave state table should not be appended to galera transaction
-    writeset.
+    We should replicate local gtid_slave_pos updates to other nodes.
+    In applier we should not append them to galera writeset.
   */
+  if (WSREP_ON_ && wsrep_thd_is_local(thd) &&
+      thd->wsrep_cs().state() != wsrep::client_state::s_none)
+  {
+    if (thd->wsrep_trx().active() == false)
+    {
+      if (thd->wsrep_next_trx_id() == WSREP_UNDEFINED_TRX_ID)
+        thd->set_query_id(next_query_id());
+      wsrep_start_transaction(thd, thd->wsrep_next_trx_id());
+    }
```

Comment:
why is new wsrep transaction started here? it would be best to piggy back gtid event inside the parent transaction

Reply:
Idea here is that if we are already executing gtid event with galera transaction, it should also mean that thd->wsrep_trx().active() is true. However, this function is also called from bg_gtid_delete_pending i.e. from background thread that periodically does cleanup and that case there might not be gtid event and wsrep transaction.